tophat/yvm
# feat: fish shell support

## Description
This resolves #116. When fish shell executes `yvm use`, it runs the bash script in bash shell, but not interactive mode. I consider running fish shell command within bash script, but every time the command is called, it boosts a new fish shell which adds a lot of overhead from the fish profile. After a few more trials, I decided to write a complete new fish shell script `yvm.fish`, which supports the same functionalities as `yvm.sh`. The main differences are:
- The added yvm path is set to `$fish_user_paths` instead because fish shell does not allow change `PATH` directly.
- Use `getNewFishUserPaths` as fish shell's path is an array, requiring a slightly different format.

The `yvm` comamnds defined in `yvm.fish` will be sourced within fish profile (`~/.config/fish/config.fish`), and it will take higher priority over functions sourced in `yvm.sh`.

The downside of this approach is that we need to maintain two copies of yvm shell script. But hopefully it will be ok since they are thin layers over the yvm logics.

## DevQA

### DevQA Prep

- Please `make install` to ensure `yvm.fish` will be copied to the YVM_DIR.
- Please use a fish shell to test this feature.

### DevQA Steps

Install YVM:
- Open a fish shell and run `make install`. 
- In another fish shell, run `type yvm`. It should say that "yvm is a function with definition" defined in the `yvm.fish` file.

Test each yvm command is working properly, including:
- install
- remove
- exec
- use
- list
- list-remote
- get-default-version
- set-default
- version
- update-self
- help

### Comments

Still misses a test for `getNewFishUserPaths`. Create the PR for preliminary review.
